### PR TITLE
Feature branch eliud koto

### PR DESCRIPTION
**Pull request by eliud-koto**
This project implements a complete machine learning workflow for credit card fraud detection. It covers data preprocessing, model training with hyperparameter tuning, experiment tracking using MLflow, evaluation, deployment through a FastAPI server, and an interactive Gradio app. Pipeline orchestration and automation are handled with Dagster.

Screenshot from 2025-10-19 15-22-59
Screenshot from 2025-10-19 15-23-19
Screenshot from 2025-10-19 15-25-49
Screenshot from 2025-10-19 15-34-40
Screenshot from 2025-10-19 15-35-35
Screenshot from 2025-10-19 15-35-52
Screenshot from 2025-10-19 15-38-32
Screenshot from 2025-10-19 15-39-44
Screenshot from 2025-10-19 15-46-25
Screenshot from 2025-10-19 15-48-12
Screenshot from 2025-10-19 15-48-52
Screenshot from 2025-10-19 15-55-17
Screenshot from 2025-10-19 15-56-08
Screenshot from 2025-10-19 15-56-20
Screenshot from 2025-10-19 15-56-30
Screenshot from 2025-10-19 15-59-24
Screenshot from 2025-10-19 16-03-33
Screenshot from 2025-10-19 16-03-53
Screenshot from 2025-10-19 16-04-00
Screenshot from 2025-10-19 16-05-45
Screenshot from 2025-10-19 16-05-58
Screenshot from 2025-10-19 16-06-13
Fraud Detection Pipeline

### CODE REVIEW

**Comment by eliud-koto** [comment]
---

# Fraud Detection Pipeline

This project builds a complete machine learning workflow to detect credit card fraud. It demonstrates data preprocessing, model training with hyperparameter tuning, experiment tracking with MLflow, evaluation, deployment via a FastAPI server, and an interactive Gradio app. Orchestration and automation are handled with Dagster.

---

## ⚙️ Model Training and Tracking

* The dataset is split into **80% training** and **20% test**.
* A **RandomForestClassifier** is trained on the training set.
* **3-fold cross-validation** is used to tune one hyperparameter: `n_estimators`.
* All experiments, parameters, and metrics are **logged using MLflow**, including nested runs for different hyperparameter values.


---

## 📊 Model Evaluation

* Evaluation is done on the **held-out 20% test set**.
* A **confusion matrix** is generated and saved as a PNG image.
* The image is logged as an **artifact in MLflow** for visual inspection.

---

## 📨 Slack Integration

After evaluation, the Dagster pipeline sends a **Slack message** to a specified channel with the model's performance metric (e.g., F1-score or accuracy).

Slack integration is handled using a simple webhook and a reusable Python function.

---

## 🌐 Model Deployment (REST API)

A **FastAPI server** is created to serve the trained model locally.

To start the server:

```bash
uvicorn serve_model:app --reload --port 8080
```

You can test it with:

```bash
curl -X POST "http://127.0.0.1:8080/predict/" \
     -H "Content-Type: application/json" \
     -d @sample_input.json
```

✅ Screenshot of successful prediction response is included in the here.

---



Screenshot from 2025-10-19 15-46-25

Screenshot from 2025-10-19 15-34-40


Screenshot from 2025-10-19 15-35-35

Screenshot from 2025-10-19 15-35-52

Screenshot from 2025-10-19 15-38-32

Screenshot from 2025-10-19 15-39-44


Screenshot from 2025-10-19 15-48-12

Screenshot from 2025-10-19 15-48-52

Screenshot from 2025-10-19 15-55-17

Screenshot from 2025-10-19 15-56-08

Screenshot from 2025-10-19 15-56-20

Screenshot from 2025-10-19 15-56-30

Screenshot from 2025-10-19 15-23-19


Screenshot from 2025-10-19 15-25-49

Screenshot from 2025-10-19 15-59-24
Screenshot from 2025-10-19 16-03-33
Screenshot from 2025-10-19 16-03-53
Screenshot from 2025-10-19 16-04-00
Screenshot from 2025-10-19 16-05-45
Screenshot from 2025-10-19 16-05-58
Screenshot from 2025-10-19 16-06-13
